### PR TITLE
Add new RunSettings Methods

### PR DESCRIPTION
**Pull request by MattToast**
Adds new `RunSettings` methods to grow the list of interface functions that are available across all `RunSettings` objects where possible. Methods added for similar arguments that are shared by two or more of the existing `RunSettings` sub-classes.

Do let me know if I am missing any additional methods for arguments that should be included, or implementations for existing `RunSettings`.

### CODE REVIEW

File: smartsim/settings/alpsSettings.py
Comment (al-rigazzi) [inline smartsim/settings/alpsSettings.py:159]: "verbosly" -> "verbosely"? (here and other times through the code)

**Review by al-rigazzi — COMMENTED** [review]
Looks good. I'd just like to understand better what was the criterion to decide which option should get a method and which not. Maybe we chat about it?
Also, I really like the warning approach for missing functions, it will help unifying driver scripts!

File: smartsim/settings/base.py
Comment (MattToast) [reply smartsim/settings/base.py]: Fixed in https://github.com/CrayLabs/SmartSim/pull/170/commits/82114ff67a5118c6ce40455fdff5519a7dcf66b6

File: smartsim/settings/alpsSettings.py
Comment (MattToast) [reply smartsim/settings/alpsSettings.py]: Fixed here and throughout code base in https://github.com/CrayLabs/SmartSim/pull/170/commits/82114ff67a5118c6ce40455fdff5519a7dcf66b6

**Comment by MattToast** [comment]
@al-rigazzi Thanks for the review!

Issue #161 goes into more detail, but basically the deciding factor for whether or not an option got made into method on the base class was whether or not is the the same or very similar functionality existed on at least two of the supported interfaces (`slurm`, `jsrun`, etc).

I did my best to try and parse docs for each of the interfaces and I think I got most of the common options (at least all the options I understood). If you can think of any other arguments I missed, do lmk and I can add them in asap!!

File: smartsim/settings/lsfSettings.py
Comment (Spartee) [inline smartsim/settings/lsfSettings.py:162]: @al-rigazzi do you think we should continue to have these aliases? or just implement the specification?

File: smartsim/settings/mpirunSettings.py
Comment (Spartee) [inline smartsim/settings/mpirunSettings.py:151]: this is a bit hard to understand. I get that you like the truthy values, but if you keep it like this, lets make sure there are some comments explaining why.

personally I feel like this would be more readable

```python
if verbose:
    ...
else:
    ...
```

File: smartsim/settings/mpirunSettings.py
Comment (Spartee) [inline smartsim/settings/mpirunSettings.py:193]: on these methods that supply a time constraint, I feel like we can improve upon "number of seconds". 

for example, what about ``hours=1, minutes=0, seconds=0`` as the default and we just convert to seconds. I feel like we could do this in other places as well.

File: smartsim/settings/mpirunSettings.py
Comment (Spartee) [reply smartsim/settings/mpirunSettings.py:151]: given this seems to be in a number of places, please either provide a reasoning why this is better then a traditional conditional or change them to a simple if/else.

File: smartsim/settings/slurmSettings.py
Comment (Spartee) [inline smartsim/settings/slurmSettings.py:225]: see comment about time

**Review by Spartee — CHANGES_REQUESTED** [review]
Good stuff! A couple things

1. there are still gaps between the implementations. ex. ``SrunSettings`` has ``set_nodes`` and few others do each class should have the *exact* same methods available that will raise warnings in the case where it's not implemented.  there should not be an ``AttributeError`` raised if I call the same methods across all classes. this is with the exception of ``JsrunSettings`` which needs weird special methods for resource sets. more on this later

2. I think we should provide a better interface for time across all the settings. I think we should be doing the translation to a string walltime. All the batch settings implement ``set_walltime`` and I think we should also provide ``set_time`` which takes in the suggestion i commented above as arguments and then formats and calls ``set_walltime`` if its a ``BatchSettings`` class.

3. @al-rigazzi why does ``AprunSettings`` randomly implement ``set_walltime``? was there a need for that? you're on the git blame.

4. Instead of ``set_timeout`` lets call it ``set_time`` and implement the suggestion above. One helper method in the base class to do the time conversion and those classes that need to can override. 

5. ``set_hostlist_from_file`` is another I think that could be implemented across all of them.

6. @al-rigazzi Is there a strict reason why we need to continue supporting the resource set semantics of LSF? can we get away with just node/task/etc if we mention that this is an alias for resource sets? I would like to get consistent.

**Comment by Spartee** [comment]
also, whitespace!!!! purge it! ha

File: smartsim/settings/mpirunSettings.py
Comment (MattToast) [reply smartsim/settings/mpirunSettings.py]: No good reason! Wasn't sure of if there where coding standards against a `if/elif` chains without a default `else` and I wanted to avoid over-indenting with a
```py
if verbose:
    ...
else:
    if "verbose" in dict:  # avoid KeyExp
        ...
```
so i opted for a guard clause instead.

Whole thing can be avoided by simply using `dict.pop`, but I am bad at my job and forgot that method exists!! Fixed in https://github.com/CrayLabs/SmartSim/pull/170/commits/fb759c0a1100c233e79014e682a3e4ca7e5b82fd

File: smartsim/settings/mpirunSettings.py
Comment (MattToast) [reply smartsim/settings/mpirunSettings.py:203]: Like the idea!! My only concern would be defaulting the `hours` param to be 1. I think it seems a bit un-intuitive that that:
```py
rs: RunSettings = ...
rs.set_time(seconds=5)
```
would set the time limit to 365 seconds from now. We could do scan the parameters and manually correct to the time to be 5 seconds seconds in the future, but I think the default being 0 / unlimited time makes more sense.

Comment (MattToast) [reply smartsim/settings/mpirunSettings.py]: Basic implementation of `set_time` in [b9aec16](https://github.com/CrayLabs/SmartSim/pull/170/commits/b9aec16118b55f342a31d4b3e77b8e60743ab137)

File: smartsim/settings/lsfSettings.py
Comment (al-rigazzi) [reply smartsim/settings/lsfSettings.py:162]: @Spartee (cc @MattToast): Answering this and your general comment: I am not really sure we can map everything in LSF to node/task/etc. We can try, and simply say that for everything else, a standard `run_arg` method can be used. 

About `set_walltime` in `AprunSettings`. I cannot think of a reason why we should have it now, I guess it made sense in an old version of the library, possibly.

File: smartsim/settings/base.py
Comment (Spartee) [inline smartsim/settings/base.py:489]: I think we should just return a string that would work with the ``env`` command here. this is overridden in every class anyway.

**Review by Spartee — COMMENTED** [review]
One small comment

File: smartsim/settings/base.py
Comment (MattToast) [reply smartsim/settings/base.py:489]: Implemented in [e887de9](https://github.com/CrayLabs/SmartSim/pull/170/commits/e887de9b7bfd74f21575428bb3ff2c213ef5f6f8)

Just want to confirm that is okay that function signature are changed by inherited classes (for instance `MpirunSettings.format_env_vars` returns a `list[str]`).

**Review by Spartee — APPROVED** [review]
Approving in advance based on conversation about unifying the return type of the ``format_env_vars`` command

File: smartsim/settings/base.py
Comment (MattToast) [reply smartsim/settings/base.py:489]: signatures unified in [2e20b85](https://github.com/CrayLabs/SmartSim/pull/170/commits/2e20b858b143dc048a2806a25b676c0892159c9b)

going to give this a final pass on monday before merging

**Pull request merged**
